allow for unique combo of columns (#235)
dbt-labs/dbt-project-evaluator: README.md

<details>
<summary><b>How to Remediate</b></summary>

> Apply a [uniqueness test](https://docs.getdbt.com/reference/resource-properties/tests#unique) and a [not null test](https://docs.getdbt.com/reference/resource-properties/tests#not_null) to the column that represents the grain of your model in its schema entry. For models that are unique across a combination of columns, we recommend adding a surrogate key column to your model, then applying these tests to that new model. See the [`surrogate_key`](https://github.com/dbt-labs/dbt-utils#surrogate_key-source) macro from dbt_utils for more info! Alternatively, you can use the [`dbt_utils.unique_combination_of_columns`](<https://github.com/dbt-labs/dbt-utils#unique_combination_of_columns-source>) test from `dbt_utils`. Check out the [overriding variables section](#overriding-variables) to read more about configuring other primary key tests for your project!

Reviewer: Can we call out here that while you can use `dbt_utils.unique_combination_of_columns`, you risk all of those columns being NULL which is not technically a primary key?

Author: I call that out [down here](https://github.com/dbt-labs/dbt-project-evaluator/pull/235/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R920)

Is that the wrong spot for it?

Reviewer: That's good with me!! Just wanted to make sure it was somewhere

Currently, this package uses different variables to adapt the models to your objectives and naming conventions. They can all be updated directly in `dbt_project.yml`

<details>
<summary><b>Testing and Documentation Variables</b></summary>

| variable    | description | default     |
| ----------- | ----------- | ----------- |
| `test_coverage_pct` | the minimum acceptable test coverage percentage | 100% |
| `documentation_coverage_pct` | the minimum acceptable documentation coverage percentage | 100% |

> | `primary_key_test_macros` | the set(s) of test macro names used to test primary key macros | [["dbt.test_unique", "dbt.test_not_null"], "dbt_utils.test_unique_combination_of_columns"] |

Reviewer: I'm confused by this "the set(s) of test macro names used to test primary key macros" - can we update it to something a bit clearer. Maybe something like "the set(s) of dbt tests used to check validity of a primary key"

1. A single test name in `project_name.test_macro_name` format
2. A list of test names in `project_name.test_macro_name` format, which will be evaluated as a group

For each entry in the parent list, the logic in `int_model_test_summary` will evaluate whether each model has all of the tests in that entry applied. If a model meets the criteria of any of the entries in the parent list, it will be considered a pass. The default behavior for this package will check for whether each model has either:

1. **Both** the `not_null` and `unique` tests applied to a single column OR
2. The `dbt_utils.unique_combination_of_columns` applied to the model. 

Therefore, the `dbt.test_unique` and `dbt.test_not_null` are grouped together in a sub-list to ensure they are evaluated together. 

*While it's not explicitly tested in this package, we strongly encourage adding a `not_null` test on each of the columns listed in the `dbt_utils.unique_combination_of_columns` tests.*

```yml
# dbt_project.yml
# set your test and doc coverage to 75% instead

> # use the dbt_constraints.test_primary_key test for fct_missing_primary_key_tests.sql

Reviewer: ```suggestion
# use the dbt_constraints.test_primary_key test to check for validity of your primary keys
```

| `documentation_coverage_pct` | the minimum acceptable documentation coverage percentage | 100% |
| `primary_key_test_macros` | the set(s) of dbt tests used to check validity of a primary key | [["dbt.test_unique", "dbt.test_not_null"], ["dbt_utils.test_unique_combination_of_columns"]] |

**Usage notes for `primary_key_test_macros:`**

The `primary_key_test_macros` variable determines how the `fct_missing_primary_key_tests` ([source](models/marts/tests/fct_missing_primary_key_tests.sql)) model evaluates whether the models in your project are properly tested for their grain. This variable is a list and each entry **must be a list of test names in `project_name.test_macro_name` format**.

For each entry in the parent list, the logic in `int_model_test_summary` will evaluate whether each model has all of the tests in that entry applied. If a model meets the criteria of any of the entries in the parent list, it will be considered a pass. The default behavior for this package will check for whether each model has either:

1. __Both__ the `not_null` and `unique` tests applied to a single column OR
2. The `dbt_utils.unique_combination_of_columns` applied to the model.

> Therefore, the `dbt.test_unique` and `dbt.test_not_null` are grouped together in a sub-list to ensure they are evaluated together.

Reviewer: Now that we're requiring a list of lists, maybe this should be updated

Author: updated this line very slightly, but if you have something else in mind let me know!

> Therefore, the `dbt.test_unique` and `dbt.test_not_null` are grouped together in a sub-list to ensure they are evaluated together.

Reviewer: ```suggestion
Therefore, each set of test(s) that define a primary key requirement must be grouped together in a sub-list to ensure they are evaluated together (e.g. [`dbt.test_unique`, `dbt.test_not_null`] ).
```

*While it's not explicitly tested in this package, we strongly encourage adding a `not_null` test on each of the columns listed in the `dbt_utils.unique_combination_of_columns` tests.*

```yml
# dbt_project.yml
# set your test and doc coverage to 75% instead
# use the dbt_constraints.test_primary_key test to check for validity of your primary keys

vars:
  dbt_project_evaluator:
    documentation_coverage_target: 75
    test_coverage_target: 75

>     primary_key_test_macros: ["dbt_constraints.test_primary_key"]

Reviewer: I believe this should now be `[["dbt_constraints.test_primary_key"]]`
